File: check_hadoop_datanode_version.pl
#!/usr/bin/perl -T
# nagios: -epn
#
#  Author: Hari Sekhon
#  Date: 2013-02-11 12:59:27 +0000 (Mon, 11 Feb 2013)
#
#  https://github.com/harisekhon/nagios-plugins
#
#  License: see accompanying LICENSE file
#

our $DESCRIPTION = "Nagios Plugin to check a Hadoop HDFS DataNode's version via NameNode JMX API

DEPRECATED - use check_hadoop_datanode_version.py which goes directly to the DataNode's JMX API, it's more efficient and easier to use as it doesn't need to specify a --node switch

Tested on Apache Hadoop 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8
";

$VERSION = "0.2";

use strict;
use warnings;
BEGIN {
    use File::Basename;
    use lib dirname(__FILE__) . "/lib";
}
use HariSekhonUtils;
use Data::Dumper;
use LWP::Simple qw/get $ua/;

set_port_default(50070);
env_creds(["HADOOP_NAMENODE", "NAMENODE", "HADOOP"], "Hadoop NameNode");

my $bean = "Hadoop:service=NameNode,name=NameNodeInfo";
my $metrics = "LiveNodes";
my $node;
my $expected;
my $list_nodes;

%options = (
    %hostoptions,
    "N|node=s"      => [ \$node,       "Node name to check" ],
    "e|expected=s"  => [ \$expected,   "Expected version (regex, optional)" ],
    "l|list-nodes"  => [ \$list_nodes, "List nodes and exit" ]
);

@usage_order = qw/host port expected list-nodes/;
get_options();

$host   = validate_host($host);
$host   = validate_resolvable($host);
$port   = validate_port($port);
my $url = "http://$host:$port/jmx?qry=Hadoop:service=NameNode,name=NameNodeInfo";
if(not defined($node)){
    usage "node not defined";
}
$expected = validate_regex($expected) if defined($expected);

vlog2;
set_timeout();
set_http_timeout($timeout - 1);

$status = "OK";

# ============================================================================ #

my $content = curl_json $url, "Hadoop daemon $host:$port";

my @beans = get_field_array("beans");

sub get_bean($){
    my $bean = shift;
    #my @beans = get_field_hash("beans");
    foreach(@beans){
        isHash($_) or quit "UNKNOWN", "invalid bean found, not a hash! $nagios_plugins_support_msg_api";
        if(get_field2($_, "name") eq $bean){
            return $_;
        }
    }
    quit "UNKNOWN", "failed to find bean with name '$bean'. $nagios_plugins_support_msg_api";
}

vlog2 "parsing JMX from '$host:$port'\n";

my $section = get_bean($bean);

my $livenodes = get_field2($section, "LiveNodes");

$json = isJson($livenodes);

if(!$json){
    quit "CRITICAL", "failed to parse json from LiveNodes metric in bean '$bean'. $nagios_plugins_support_msg_api";
}

vlog3 Dumper($json);

if($list_nodes){
    print("DataNodes:\n\n");
    foreach(sort keys %{$json}){
        print("$_\n");
    }
    exit $ERRORS{"UNKNOWN"};
}

my @nodes = grep { $_ =~ /^$node(?::\d+)?$/ } keys %{$json};
if(not @nodes){
    quit "CRITICAL", "datanode '$node' not found in NameNode JMX";
}
if(scalar @nodes > 1){
    quit "CRITICAL", "more than one node matched!! $nagios_plugins_support_msg";
}

my $version = get_field2($json->{$nodes[0]}, "version");

if(!isVersion($version)){
    quit "CRITICAL", "unrecognized version '$version'. Format may have changed. $nagios_plugins_support_msg";
}

$msg = "Hadoop DataNode '$node' version = $version";
check_regex($version, $expected) if defined($expected);

quit $status, $msg;

File: check_hadoop_datanode_version.py
#!/usr/bin/env python
#  coding=utf-8
#  vim:ts=4:sts=4:sw=4:et
#
#  Author: Hari Sekhon
#  Date: 2017-09-06 13:09:35 +0200 (Wed, 06 Sep 2017)
#
#  https://github.com/harisekhon/nagios-plugins
#
#  License: see accompanying Hari Sekhon LICENSE file
#
#  If you're using my code you're welcome to connect with me on LinkedIn
#  and optionally send me feedback to help steer this or other code I publish
#
#  https://www.linkedin.com/in/harisekhon
#

"""

Nagios Plugin to check the version of the Hadoop DataNode via the JMX API

Set --port to 1022 for Kerberized clusters

Tested on HDP 2.6.1 and Apache Hadoop 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8

"""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import os
import sys
import traceback
srcdir = os.path.abspath(os.path.dirname(__file__))
libdir = os.path.join(srcdir, 'pylib')
sys.path.append(libdir)
try:
    # pylint: disable=wrong-import-position
    from harisekhon.utils import UnknownError
    from harisekhon import RestVersionNagiosPlugin
except ImportError as _:
    print(traceback.format_exc(), end='')
    sys.exit(4)

__author__ = 'Hari Sekhon'
__version__ = '0.2'


# pylint: disable=too-few-public-methods
class CheckHadoopDataNodeVersion(RestVersionNagiosPlugin):

    def __init__(self):
        # Python 2.x
        super(CheckHadoopDataNodeVersion, self).__init__()
        # Python 3.x
        # super().__init__()
        self.name = 'Hadoop DataNode'
        self.default_port = 50075
        self.path = '/jmx?qry=Hadoop:service=DataNode,name=DataNodeInfo'
        self.json = True
        self.auth = False
        self.ok()

    # must override, cannot change to @staticmethod
    def parse_json(self, json_data):  # pylint: disable=no-self-use
        data = json_data['beans'][0]
        if data['name'] != 'Hadoop:service=DataNode,name=DataNodeInfo':
            raise UnknownError('name {} != Hadoop:service=DataNode,name=DataNodeInfo'.format(data['name']))
        version = data['Version']
        #log.info("raw version = '%s'", version)
        version = version.split(',')[0]
        return version


if __name__ == '__main__':
    CheckHadoopDataNodeVersion().main()
